### ISSUE TITLE
Change Alert Keys

### ISSUE THREAD

**Issue by ahmedbodi** [bitcoinclassic/bitcoinclassic#27]
Wont be changed in first release. 
May be changed in future releases.

**Comment by dgmib** [comment]
I purpose that instead of changing the existing alert key, we add support for multiple alert keys and validate alerts against all known keys.  In the UI & logs, we can display who the owner of the key (and therefore sender of the alert is).  Alerts sent with the original key will still be relayed unaltered and uncensored, but it'd be clear who sent them.

Proof of concept code can be seen here: https://github.com/dgmib/bitcoinclassic/pull/1

**Comment by ripper234** [comment]
ACK

Is there a consider.it page for this?

**Comment by jtoomim** [comment]
I think this issue deserves consideration, and is related enough to the concept of a hard fork to merit discussion. In particular, the Alert system may be useful in the deployment of hard forks, and may be used in opposition to the hard fork.

However, I think we need a description of the likely scenarios first, and how they would play out, before we start getting votes. Let's make sure we understand everything. Yes?

**Comment by ghost** [comment]
Alert Key change is a must in dual/plural bitcoin client development atmospheres. Specific per client. Alert keys are meant for public consumption of information per community fork. Fork meaning "bitcoin" itself based on the majority.

This is not meant to be a "Whoever holds the key, gets to broadcast an emergency message across all clients" scenario.

**Comment by Hannott** [comment]
If I may.
A very short risk analysis:
Case: Version 0x20000008 blocks mined start to go high, pushing 50%.
Risk: Theymos, Luke-Jr, GMaxwell or others use current alert key to discredit Bitcoin Classic.
Chance: Pretty fucking high, in my opinion. Most of all from Theymos.
Damage: Miners are afraid of FUD and could revert to Core.

ACK, and this is a blocker imo.

**Comment by ripper234** [comment]
+1 to blocker

**Comment by dgmib** [comment]
I'm with Jon on this, we need to first come up with some well thought out solutions before it goes to consider.it for voting.

I've started a post on reddit to discuss this here:

https://www.reddit.com/r/Bitcoin_Classic/comments/41jxgi/request_for_comments_changing_bitcoins_alert_keys/

BitcoinClassic has gathered momentum by doing one thing better than "the other guys": Listening to the users.  Let's keep doing that before we put this to a vote.

**Comment by andyrowe** [comment]
Pseudo-ACK, +1 that it's a blocker.

I think the best solution is to leave the original alert keys, but code a backup set of alert keys that activate only if the original alert keys are nulled.  Have an option in case old key holders go nuclear, but leave them the rope to hang themselves with while also having an out.

**Comment by ripper234** [comment]
KISS.

Remove original alert keys, they are no longer needed or relevant.

On Tue, Jan 19, 2016 at 1:35 AM, Andy Rowe notifications@github.com wrote:

> Pseudo-ACK, +1 that it's a blocker.
> 
> I think the best solution is to leave the original alert keys, but code a
> backup set of alert keys that activate only if the original alert keys are
> nulled. Have an option in case old key holders go nuclear, but leave them
> the rope to hang themselves with while also having an out.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/bitcoinclassic/bitcoinclassic/issues/27#issuecomment-172681863
> .

## 

Ron Gross
Skype = ripper234, Schedule my time - meetme.so/RonGross
CNO http://www.chiefninjaofficer.com/

**Comment by PartTimeLegend** [comment]
I propose a shared key system. Say we assign an odd number of key holders, say 5 for an arbitrary figure. We must have consensus between a minimum of 3 to send an alert.

Alternatively all key holders are given different keys, so they can be revoked without the need for a fork.

**Comment by ripper234** [comment]
That's a great idea - but out of scope here.
KISS - first change the alert keys.

In future releases we can implement more sophisticated methods.

On Tue, Jan 19, 2016 at 12:31 PM, Antony Bailey notifications@github.com
wrote:

> I propose a shared key system. Say we assign an odd number of key holders,
> say 5 for an arbitrary figure. We must have consensus between a minimum of
> 3 to send an alert.
> 
> Alternatively all key holders are given different keys, so they can be
> revoked without the need for a fork.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/bitcoinclassic/bitcoinclassic/issues/27#issuecomment-172811542
> .

## 

Ron Gross
Skype = ripper234, Schedule my time - meetme.so/RonGross
CNO http://www.chiefninjaofficer.com/

**Comment by BitcoinXio** [comment]
There are a few good ideas on this thread. @Hannott's risk assessment is spot on, and it's without any doubt that theymos will issue an alert post-fork condemning Bitcoin Classic. 

There is also a ConsiderIt already with overwhelming agreement that theymos' alert keys should be removed: https://bitcoinclassic.consider.it/remove-theymos-alert-keys-access

@PartTimeLegend has a good suggestion to issue shared keys. It's trivial for Classic to issue new keys and have a number set aside for trusted community members from the old 'Core' client, for example, if one day Satoshi were to out of nowhere decide to become public. If he were to cryptographically sign an old bitcoin address to prove that it is him, we could issue him a key, etc. Just because new keys are issued doesn't mean people from Core can't be issued Alert keys; but the point is to remove any threat from those that hold them now, which we all undoubtedly know will happen.

**Comment by jtoomim** [comment]
I think people are failing to explore what happens at the protocol level when Bitcoin Classic nodes reject Alert messages sent by other nodes. Can someone look into the peer banning code and see if this would cause a network split?

Also, I believe the alert keys allow us to terminate a message as well as start one, and can also put the alert system into a DoS mode. Can someone please look up the details of how the alert system works, and maybe post a link?

**Comment by dgmib** [comment]
Yes, I've looked into this.  It doesn't happen immediately, but **peer banning will kick in if enough alerts signed by an unrecognized key come through.**

**If classic doesn't relay alerts signed by core it will cause network issues.**

**Comment by ChainQuery** [comment]
https://en.bitcoin.it/wiki/Alert_system
Alerts contain: ID, Sent date, Expires (UTC), Versions, Priority, Message 
Doesn't @gavinandresen have the alert key as well?

**Comment by jtoomim** [comment]
The problem is that there is no "theymos's key", there is just THE alert key for all of the Bitcoin clients. If we remove the alert key, then we can no longer interpret alert messages sent to other Bitcoin clients, and we can no longer send alert messages that will reach both sets of Bitcoin clients.

I'd like to talk with Gavin about this in more detail, since he's the one who actually has the alert key and has used it in the past. However, the last time I talked to him about this (and Jeff, IIRC), they were not in favor of it.

I agree that it's a sticky situation. However, I think that removing the one and only Alert key is likely to be worse than keeping it in. We don't really care too much if Theymos sends nasty messages to Classic clients, do we? We care about what he sends to Core clients, and we can't do anything about that except let people know that they should expect it.

@dgmib, I edited your comment to add some bold to the formatting.

It's also worth mentioning that adding alert keys to Classic and sending unrecognized alert messages to Core would also cause peer banning. The problem happens in both ways.

**Comment by dgmib** [comment]
> It's also worth mentioning that adding alert keys to Classic and sending unrecognized alert messages to Core would also cause peer banning. The problem happens in both ways.

This is exactly what I'm worried about. Changing the alert key might not fork the block chain, but it needs to be treated like a hard fork change.  Everybody needs to upgrade or we risk breaking the alert system, and possible the entire network.

I know people don't like that Themos has the alert key, but I think the risks of making changes are far greater than any damage Themos could do.  Themes might be power drunk, but he's not stupid.  Sending an "FUD" alert, would only end up badly for him in the end.

**Comment by jtoomim** [comment]
@BitcoinXio: I consider https://bitcoinclassic.consider.it/remove-theymos-alert-keys-access to not describe this issue correctly, and do not consider the vote there to be relevant. 

The proposal in this github issue is to remove the _only_ alert key for Bitcoin.

The proposal on consider.it is to remove _theymos's_ alert key for Bitcoin. This is inaccurate, as it implies that he has an alert key all to himself. He doesn't.

I think that's a pretty big inaccuracy, and so I will not consider that vote to have standing on this issue.

The consider.it page also is missing important technical details about how the alert system works, like dgmib's comment above. This may also be sufficient grounds for a vote restart.

I'm going to try to get back to version bits code. I'll try to check back on this thread in a day or so.

**Comment by ChainQuery** [comment]
Interestingly, and possibly problematically the wiki states:

> The known private key holders are Satoshi Nakamoto, Gavin Andresen and theymos. There are other people able to issue alerts in the event of the incapacitation of the aforementioned.

Who are the "other people"? I don't believe there is a way to identify who published an alert...

**Comment by dgmib** [comment]
@ChainQuery: You're correct, it's not possible to identify the publisher of the alert without network wide changes.

**Comment by dgmib** [comment]
It's also not possible to know for sure who all has the alert key either.  Gavin may know of other people.  But there's no way to know for certain if it's an exhaustive list.

**Comment by gavinandresen** [comment]
Wladimir also has the alert key, and has used it once or twice since I stepped down as lead maintainer for Core.

I think we should assume that none of the existing key-holders will be irresponsible; if they are, any keyholder can send a "final alert" that overrides any other alerts and simply displays the message:

URGENT: Alert key compromised, upgrade required

... to all versions of the software (Core and Classic).

Replacing the alert system has been on Core's TODO list for a long time, but it has never been high enough priority to get done.

I don't think improving or replacing the alert system is a blocker for the first release of Classic.

**Comment by dgmib** [comment]
I agree with gavin, no changes to the alert system before the first release of Classic.

@gavinandresen: Has there been any discussion among the core devs on what they want to replace the alert system with?

**Comment by ChainQuery** [comment]
Agreed w/ Gavin shelve any alert system changes for a later release.

**Comment by jtoomim** [comment]
Closing for now. We should reopen this later, but for now it's causing too much of a distraction.

**Comment by ripper234** [comment]
@jtoomim instead of closing, you should tag this something like "not-for-this-release" and keep it open.

**Comment by ripper234** [comment]
:+1:

**Comment by 123lee** [comment]
> @dgmib **if classic doesn't relay alerts signed by core it will cause network issues.**

Can you explain how this could happen? Would core client's know that a classic client is not sending a signed alert (when it should be) and 'ban' it?

If that is not the case, can we not simply deactivate the alert system. Is it really of any use any more?

**Comment by christophebiocca** [comment]
@123lee:

There's two changes that can lead to inter-node bans:
1. We add a new alert key. When forwarding our alerts, core nodes see it as invalid, and increase our DoS count (which can eventually lead to a ban).
2. We remove the old alert key. When core forwards alerts, our nodes see it as invalid, and increase their DoS count (which can eventually lead to a ban).

We can fix 2 by special-casing the old key as "invalid but not DoS".

**Comment by dgmib** [comment]
@christophebiocca is correct.  If we add a new key, and send an alert with it we will "fork" the network into classic nodes and core nodes as core node will automatically start banning classic nodes.  We can make a special case so that we don't forward alerts with the old key but don't ban them either which would make the existing alert system unreliable but not totally broken.  But since any alert key holder (including Gavin) already has a way to quickly and reliably disable the alert system network wide (if there should be a reason to), there's no good reason to pre-emptively remove or disable the old key.

**Comment by tymat** [comment]
ACK

**Comment by zander** [comment]
Core removed this feature in total. Future releases of Core will no longer show these notifications.

As long as there are plenty of old Core nodes around, any inaction on our part is not really an issue either way.  But I do think that after a hardfork (which implies everyone will stop using the old core releases and get something new) we should consider if it makes sense to have an alert system which the lead devs of a competing implementation can trigger.

Is there any reason why having such an alert system makes sense?

**Comment by gavinandresen** [comment]
The alert system is great for 'unknown unknowns', like a vulnerability in network-facing code (which could be in code we don't directly control-- e.g. bug in boost or compiler bug).

I think it is worth keeping, with a fresh alert key.

**Comment by dgmib** [comment]
@gavinandresen If we setup a new alert key, we'll need to also do something so we don't send alerts to existing core nodes and get ourselves peer banned.  Either a parallel alert system, or just some logic that doesn't send alerts to non-classic nodes.  Do you have a recommendation, on how to best implement that?

Also we need to keep in mind, that if non-classic nodes aren't relaying these alerts.  Then some classic nodes likely won't get the alerts. Unless classic becomes the majority implementation, there will likely be nodes that purely out of coincidence don't have any classic peers.

**Comment by gavinandresen** [comment]
How to send only to classic nodes: only send alerts to peers with "classic"
in the right place in their user-agent.

What if a node has no classic peers: lets not let the perfect be the enemy
of the good. If necessary, we could write alert-o-matic code that connects
and pushes alerts (and then disconnects) from classic nodes that it finds
from the public data that some of the DNS seeds publish.  Or we could run a
classic-only DNS seeder and always get at least one peer from it...

**Comment by saltduck** [comment]
If we set a new nVersion for classic node, we can only send alert to classic nodes only. Am I right?

**Comment by PartTimeLegend** [comment]
As alert keys are removed now, can we close this?

**Comment by zander** [comment]
> As alert keys are removed now, can we close this?

Alert keys have not been removed in Classic.

**Comment by gmaxwell** [comment]
Please either remove the alert system or change the keys. I must insist that you do this immediately. Your project's continual dishonest blaming of Bitcoin Core for your own errors and ignorance demands the minimal possible entanglement.

I would appreciate it if you provided a clear plan with time-frames for this update.

**Comment by gmaxwell** [comment]
@zander Please acknowledge receipt of the above message and advise when this work will be performed.

**Comment by Hannott** [comment]
@gmaxwell Could you be more of an asshole about it? That would probably speed up the process for sure! /s

**Comment by jl2012** [comment]
Since many people share the same key, which not all of them are Bitcoin Classic developers, it is very risky for this project to keep using the key. It is also very unfair to some key holders since they may not want to hold a critical key of a project that they do not involve.

And this is the development standard of Bitcoin Classic:
Q: Will you fix an obvious DoS vulnerability?
A: We will consider.it™

**Comment by TheBlueMatt** [comment]
Might I remind folks that the alert key was sent around to a bunch of people, including
gpg: encrypted with 4096-bit RSA key, ID 8864A82BEACB3C76, created 2010-07-22
      "Mark Karpelès mark@hell.ne.jp"
and several PGP keys which are now marked as having been compromised? I would say that, at this point, sending the "URGENT: Alert key compromised, upgrade required" message around to all nodes would be prudent.

**Comment by zander** [comment]
I just write in to point out that GitHub is not a discussion forum, it literally is one of the worst platforms you can choose for that.

There already is a consider.it page which is a better platform, although I welcome community members opening other places for discussion to find best solutions and come back with a merge request.

For completeness-sake I will repeat the consider.it link here; https://bitcoinclassic.consider.it/remove-theymos-alert-keys-access

**Comment by jl2012** [comment]
@zander calling it "theymos alert key" when it is well known that many people share the key is tragically dishonest. This is nothing but a purely social attack.

**Comment by gmaxwell** [comment]
@zander I'm not asking for a discussion. I'm asking for a timeframe for when the work will be completed and released.

The PR that removed it completely in Bitcoin upstream was https://github.com/bitcoin/bitcoin/pull/7692

**Comment by Hannott** [comment]
@gmaxwell Oh, you have authority on this matter? It will or won't get done when _not you_ decides.

**Comment by ntpd** [comment]
@gavinandresen @zander The solution is fairly obviously to change the message type field to _any other string other than alert_, and then you won't relay them to anybody (messages of an unknown type are dropped by the node). Stop making this harder than it needs to be, it can literally be a single byte change.

**Comment by ntpd** [comment]
Here's a patch that both changes the message to only be relayed to Bitcoin Classic peers without hitting old versions that will ban you for using the wrong message start, and also updates the alert key to be the one from Sunny King's PeerCoin repository (hope you don't mind @gavinandresen!) 

```
diff --git a/src/chainparams.cpp b/src/chainparams.cpp
index d4ee247..8a97973 100644
--- a/src/chainparams.cpp
+++ b/src/chainparams.cpp
@@ -101,7 +101,7 @@ public:
         pchMessageStart[1] = 0xbe;
         pchMessageStart[2] = 0xb4;
         pchMessageStart[3] = 0xd9;
-        vAlertPubKey = ParseHex("04fc9702847840aaf195de8442ebecedf5b095cdbb9bc716bda9110971b28a49e0ead8564ff0db22209e0374782c093bb899692d524e9d6a6956e7c5ecbcd68284");
+        vAlertPubKey = ParseHex("04a0a849dd49b113d3179a332dd77715c43be4d0076e2f19e66de23dd707e56630f792f298dfd209bf042bb3561f4af6983f3d81e439737ab0bf7f898fecd21aab");
         nDefaultPort = 8333;
         nMaxTipAge = 24 * 60 * 60;
         nPruneAfterHeight = 100000;
diff --git a/src/protocol.cpp b/src/protocol.cpp
index 36e31e7..d8c4123 100644
--- a/src/protocol.cpp
+++ b/src/protocol.cpp
@@ -28,7 +28,7 @@ const char *GETADDR="getaddr";
 const char *MEMPOOL="mempool";
 const char *PING="ping";
 const char *PONG="pong";
-const char *ALERT="alert";
+const char *ALERT="benevolent";
 const char *NOTFOUND="notfound";
 const char *FILTERLOAD="filterload";
 const char *FILTERADD="filteradd";
```

**Comment by ahmedbodi** [comment]
swap alert for alert-classic maybe?

**Comment by ntpd** [comment]
@ahmedbodi The message type is a null padded char[12], alertclassic would fit.

**Comment by gmaxwell** [comment]
Thank you. Do you have a planned date for a release with this update?

**Comment by gmaxwell** [comment]
A release is better than a planned date. Thanks.

**Comment by NiKiZe** [comment]
Greg stopp trolling, a release has already been done!
